Support for edits

dkuegler opened the pull request:
This PR adds support for edits in FastSurfer.

Edits are enabled by the `--edits` flag for `run_fastsurfer.sh`.

Edits supported:
<asegdkt_segfile> via <asegdkt_segfile-w/o-ext>.manedit.<ext>
white matter segmentation (wm.mgz) via wm.manedit.mgz
white matter control points (via the standard FreeSurfer control points file)
talairach registration via mri/mri/transforms/talairach.xfm (overwriting talairach registration via mri/mri/transforms/talairach.xfm)
brain.finalsurfs.mgz via brain.finalsurfs.manedit.mgz


This PR also move the allow_root check into functions.sh and introduces run_it and run_it_cmdf function variants that are more clear wrt. argument splitting (see SC2206, SC2207 and similar).

m-reuter commented:
Hi, thanks for the pr. It would be great to fix the spelling errors and also add a readme for edits, which can be pretty short in the first iteration, with a 1-2 sentence description of each possible edit.

dkuegler commented:
Spelling should be fixed, documentation will need to be a future thing for now.

m-reuter commented on recon_surf/recon-surf.sh:
```diff
@@ -1002,36 +1044,39 @@ for hemi in lh rh ; do
 
   # these are run automatically in fs7* recon-all and cannot be called directly without -pial flag (or other t2 flags)
   # they are the same for fsaparc and long
-  echo "pushd $mdir > /dev/null" >> "$CMDF"
```
Have you tested this? I am not sure why it was cd into mri dir, but maybe there was a reason. With your change it is now running from within surf dir which might change things.

dkuegler replied:
I had tested some of these, but this clearly is a bug :-(

m-reuter commented on recon_surf/talairach-reg.sh:
```diff
@@ -55,60 +88,113 @@ else
-  RunIt "${cmd[*]}" "$LF"
+pushd "$mdir/transforms" > /dev/null || ( echo "ERROR: Could not change to the transforms directory $mdir/transforms!" | tee -a "$LF" ; exit 1 )
+  if [[ ! -e talairach_with_skull.lta ]] ; then
+    cmd=(ln -sf talairach.xfm.lta talairach_with_skull.lta)
```
shouldn' these be ln_or_copy ?

dkuegler replied:
It should, I am not sure why, but I believe this has never changed to `softlink_or_copy`

m-reuter commented on run_fastsurfer.sh:
```diff
@@ -220,6 +227,9 @@ SURFACE PIPELINE:
                             etiv estimates for 3T MR images, default: 1.5T atlas).
+  --edits                 Disable the check for existing recon-surf.sh run and enables
+                            edits of mri/wm.mgz and brain.finalsurfs.mgz by manedit
+                            files as well as FreeSurfer-style WM control points.
```
I think it is strange to have the flag --edits in the help text at multiple places. I think it should be a single help block explaining all of its effects in one location.

dkuegler replied:
I guess consolidating this into the first text makes more sense. On hindsight, I agree. I think --no_biasfield also has multiple mentions, so that should also be changed.

dkuegler commented:
This branch is currently a mess of fixes and features for 6 different things. Those need to be separated into individual branches and PRs.

At a minimum, this version seems to work for longitudinal processing (still unclear for edits).

dkuegler committed:
run_fastsurfer.sh
- Adds the --edits flag
- Add additional log messages for edits
- Support for manedit files for asegdkt_segfile.
- Also propagate asegdkt_segfile into aseg_segfile and mask_name

recon_surf/recon-surf.sh
- Adds the --edits flag
- Add additional log messages for edits
- Change check for existing run into a message instead a fail criteria in edits mode
- Do not delete logfile in edits mode
- Make mask and aseg_noCC required for edits mode (otherwise it might get unclear what should or was already done between run_fastsurfer and recon-surf)
- Add manedit "search" for mask and asegdkt_segfile.

recon_surf/talairach-reg.sh
- Add an extra argument (new argument #3 to indicate whether talairach.reg should support edits)
- Support for edits: some messages and if in edit mode, only run talairach registration if mri/transforms/talairach.xfm and mri/transforms/talairach.auto.xfm are the same (so there are no edits) -- thus edits in talairach.xfm superceed automatic results
- talairch-reg.sh will now fail if mri/transforms/talairach.xfm exists, but the script is not running in edits mode (independent if talairach.auto.xfm is the same or not).

recon_surf/functions.sh
- add the add_file_suffix function, which adds a file suffix (for example manedits) to a file name

m-reuter commented:
I rebased this to current dev which disentangles the edits from the other changes. Please check :
- that everything is as wanted and
- no purposeful changes from the earlier commits are undone

Also update your local with a hard reset.

dkuegler committed:
Formatting of comments

dkuegler committed:
Update and rewrite the editing documentation to cover all edits that are now possible.

m-reuter committed:
Update EDITING.md

dkuegler committed:
Move FreeSurfer equivalents to a guide at the back
Add what NOT to do section

neginshirvani committed:
Review and edit the documentation

dkuegler committed:
Second review pass of the documentation

dkuegler commented:
Maybe there should be a check if the fastsurfer versions are the same between the original run and the rerun? @m-reuter 

If this is just a warning or a message this will likely get ignored.
If this is a failure, it might get more annoying than not.
Solution might be to write a message to the log and let the user confirm it (enter).
But if nothing is attached to stdin, the process proceeds no matter what so only interactive sessions stop (this is how install scripts sometimes solve this)...

m-reuter commented:
> Maybe there should be a check if the fastsurfer versions are the same between the original run and the rerun?

I think warning message in the logs is fine. It should only stop if versions are not compatible any longer (and it would fail somewhere during processing). Otherwise, if someone edits things, that is changing more than most of our versions.

m-reuter commented on doc/overview/EDITING.md:
````diff
@@ -1,125 +1,213 @@
+```
+
+Note, a re-run of the segmentation pipeline, as in the command above, should not be harmful, but is only required if the [asegdkt_segfile](#asegdkt_segfile) was edited.
+There, we can usually skip the segmentation step with
````
Maybe you mean: "Therefore, in most cases, we can skip ..."

dkuegler replied:
I think you actually wrote this sentence :).

```suggestion
Therefore, in most cases, we can skip the segmentation step with
```

m-reuter commented on doc/overview/EDITING.md:
```diff
@@ -1,125 +1,213 @@
+### When to use
+Over- and/or under-segmentation of the cortical gray matter: voxels that should be gray matter are excluded, or those that should not are included.
+
+This explicitly 
```
Sentence is missing.

dkuegler replied:
Also just remove, I think the decision was to keep the docs of these "special edits" a bit shorter.
```suggestion
```

m-reuter commented on doc/overview/EDITING.md:
```diff
@@ -1,125 +1,213 @@
+Often, these errors should be fixed in [asegdkt_segfile](#asegdkt_segfile) `<subject_dir>/mri/aparc.DKTatlas+aseg.deep.manedit.mgz`, but if that is not successful:
+1. Open end edit `<subject_dir>/mri/brain.finalsurfs.manedit.mgz` (overwriting values in `<subject_dir>/mri/brain.finalsurfs.mgz`).
+2. [Re-run FastSurfer](#general-process) to fix the pial surface. 
+3. 
```
item 3. is missing

dkuegler replied:
I think this should just be removed... The docs here does not really have a step, it would just be to check if the results are now better :)
```suggestion
```

m-reuter commented on doc/overview/EDITING.md:
```diff
@@ -1,125 +1,213 @@
+3. Brain extraction by editing `<subject_dir>/mri/brain.mgz` (delta to `brain.auto.mgz`) -> [asegdkt_segfile](#asegdkt_segfile)
+4. Subcortical segmentation: `<subject_dir>/mri/aseg.presurf.mgz` (delta to `aseg.presurf.auto.mgz`) -> [asegdkt_segfile](#asegdkt_segfile)
+5. CRS seeds: PONS, CC, LH, RH, watershed => seed files (`<subject_dir>/scripts/seed-(pons|cc|lh|rh|ws).crs.man.dat`) -> [asegdkt_segfile](#asegdkt_segfile)
+6. WM control points: `<subject_dir>/tmp/control.dat` -> [asegdkt_segfile](#asegdkt_segfile)
```
hm, we could allow them to create control points which should be used by recon-surf when crating wm.mgz?

dkuegler replied:
I think we could, but there are some comments somewhere why we would not want to do that. I think I did not really understand those comment but just assumed you had decided against this on principle.

dkuegler replied:
Also, I am not really sure how these labels would propagate back into aseg.mgz and aparc.DKTatlas+aseg.mgz

m-reuter replied:
They would not, they are only used to increase intensity in thin WM regions to avoid loosing a gyrus or parts of a lobe (usually temporal lobe) due to low grey white contrast there. Let's discuss.

m-reuter committed:
Update doc/overview/EDITING.md

Co-authored-by: David Kügler <david.kuegler@dzne.de>

m-reuter commented on recon_surf/recon-surf.sh:
```diff
@@ -87,6 +87,12 @@ FLAGS:
                             that segmentations produced otherwise are conformed.
                             Requires an ABSOLUTE Path! Default location:
                             \$SUBJECTS_DIR/\$sid/mri/aparc.DKTatlas+aseg.deep.mgz
+  --mask_name <mask_file> Path to the brainmask file to use. Default location:
```
```suggestion
  --mask_name <mask_file> Path to the brain mask file to use. Default location:
```

m-reuter committed:
Update recon_surf/recon-surf.sh
